File: backend/services/admin/src/Enma.Admin.Application/Services/OrganizationMembersService.cs
using Enma.Admin.Application.Abstractions;
using Enma.Admin.Application.Contracts;
using Enma.Admin.Application.Dto.Notifications;
using Enma.Admin.Application.Dto.OrganizationMembers;
using Enma.Common.Enums;
using FluentResults;

namespace Enma.Admin.Application.Services;

internal sealed class OrganizationMembersService : IOrganizationMembersService
{
    private readonly IOrganizationMembersRepository _organizationMembersRepository;
    private readonly INotificationsService _notificationsService;

    public OrganizationMembersService(
        IOrganizationMembersRepository organizationMembersRepository,
        INotificationsService notificationsService)
    {
        _organizationMembersRepository = organizationMembersRepository;
        _notificationsService = notificationsService;
    }

    public async Task<Result<OrganizationMemberDto>> GetAsync(
        Guid orgId,
        Guid userId,
        CancellationToken ct = default)
    {
        var res = await _organizationMembersRepository.GetAsync(orgId, userId, ct);
        return res.IsSuccess
            ? Result.Ok(res.Value.ToDto())
            : Result.Fail<OrganizationMemberDto>(res.Errors);
    }

    public Task<Result<bool>> IsMemberAsync(Guid orgId, Guid userId, CancellationToken ct = default)
        => _organizationMembersRepository.IsMemberAsync(orgId, userId, ct);

    public async Task<Result<IReadOnlyList<OrganizationMemberDto>>> ListByOrgAsync(
        Guid orgId,
        int offset,
        int limit,
        CancellationToken ct = default)
    {
        var res = await _organizationMembersRepository.ListByOrgAsync(orgId, offset, limit, ct);
        return res.IsSuccess
            ? Result.Ok<IReadOnlyList<OrganizationMemberDto>>(res.Value.Select(x => x.ToDto()).ToList())
            : Result.Fail<IReadOnlyList<OrganizationMemberDto>>(res.Errors);
    }

    public async Task<Result> SetRoleAsync(
        Guid orgId,
        Guid userId,
        SetOrganizationMemberRoleDto dto,
        CancellationToken ct = default)
    {
        var res = await _organizationMembersRepository.SetRoleAsync(orgId, userId, dto.Role, ct);
        if (res.IsFailed)
        {
            return Result.Fail(res.Errors);
        }

        await _notificationsService.CreateAsync(new CreateNotificationDto(
            RecipientUserId: userId,
            Type: NotificationType.OrganizationRoleChanged,
            Title: "Organization role changed",
            Message: $"Your organization role has been changed to {dto.Role}.",
            ResourceId: orgId), ct);

        return Result.Ok();
    }

    public async Task<Result> SetStatusAsync(
        Guid orgId,
        Guid userId,
        SetOrganizationMemberStatusDto dto,
        CancellationToken ct = default)
    {
        var res = await _organizationMembersRepository.SetStatusAsync(orgId, userId, dto.Status, ct);
        return res.IsSuccess
            ? Result.Ok()
            : Result.Fail(res.Errors);
    }

    public async Task<Result> RemoveAsync(Guid orgId, Guid userId, CancellationToken ct = default)
    {
        var res = await _organizationMembersRepository.RemoveAsync(orgId, userId, ct);
        if (res.IsFailed)
        {
            return Result.Fail(res.Errors);
        }

        await _notificationsService.CreateAsync(new CreateNotificationDto(
            RecipientUserId: userId,
            Type: NotificationType.RemovedFromOrganization,
            Title: "Removed from organization",
            Message: "You have been removed from an organization.",
            ResourceId: orgId), ct);

        return Result.Ok();
    }
}

File: frontend/src/services/admin/organizationMembersService.ts
import { apiClient } from "@/api/apiClient";
import type {
  Guid,
  OrganizationMemberDto,
  SetOrganizationMemberRoleDto,
  SetOrganizationMemberStatusDto,
} from "@/types/admin.types";

export default class OrganizationMembersService {
  private orgBaseUrl(organizationId: Guid): string {
    return `/api/admin/v1/organizations/${organizationId}/members`;
  }

  public async get(organizationId: Guid, userId: Guid): Promise<OrganizationMemberDto> {
    const { data } = await apiClient.get<OrganizationMemberDto>(`${this.orgBaseUrl(organizationId)}/${userId}`);
    return data;
  }

  public async isMember(organizationId: Guid, userId: Guid): Promise<boolean> {
    const { data } = await apiClient.get<boolean>(`${this.orgBaseUrl(organizationId)}/${userId}/is-member`);
    return data;
  }

  public async list(organizationId: Guid, offset = 0, limit = 50): Promise<OrganizationMemberDto[]> {
    const { data } = await apiClient.get<OrganizationMemberDto[]>(this.orgBaseUrl(organizationId), {
      params: { offset, limit },
    });
    return data;
  }

  public async setRole(organizationId: Guid, userId: Guid, dto: SetOrganizationMemberRoleDto): Promise<void> {
    await apiClient.patch(`${this.orgBaseUrl(organizationId)}/${userId}/role`, dto);
  }

  public async setStatus(organizationId: Guid, userId: Guid, dto: SetOrganizationMemberStatusDto): Promise<void> {
    await apiClient.patch(`${this.orgBaseUrl(organizationId)}/${userId}/status`, dto);
  }

  public async remove(organizationId: Guid, userId: Guid): Promise<void> {
    await apiClient.delete(`${this.orgBaseUrl(organizationId)}/${userId}`);
  }
}
